mlir/include/mlir/Dialect/GPU/IR/GPUOps.td

```diff
@@ -1345,12 +1345,40 @@ def GPU_BarrierOp : GPU_Op<"barrier"> {
   let hasCanonicalizer = 1;
 }
 
+def SpatialExtentsAttr
+  : GPU_Attr< "SpatialExtents", "spatial_extents"> {
+  let summary = "Specify the spatial extent of a gpu module such as subgroup size.";
+  let description = [{
+    Specify the spatial extent of a GPUModuleOp. This can be useful for enabling
+    algorithms, optimizations, or features that rely on particular kernel
+    parameters being fixed, e.g., vendor-specific extentions that only work for
+    one subgroup size. `reqdSubgroupSize` specifies that the module is only
+    valid when subgroup size is that value. `reqdWorkgroupSize` specifies that
+    the module is only valid when workgroup size is that value.
+    `maxWorkgroupSize` specifies that the module is only valid when workgroup
+    size is less than or equal to that value.
+
+    All parameters are optional. Having both `reqdWorkgroupSize` and
+    `maxWorkgroupsize` is invalid.
+  }];
+
+  let parameters = (ins
+  OptionalParameter<"IntegerAttr">: $reqdSubgroupSize,
+  OptionalParameter<"DenseI32ArrayAttr">: $reqdWorkgroupSize,
+  OptionalParameter<"DenseI32ArrayAttr">: $maxWorkgroupSize);
+
+  let assemblyFormat = "`<` struct(params) `>`";
+
+  let genVerifyDecl = 1;
+}
+
 def GPU_GPUModuleOp : GPU_Op<"module", [
       DataLayoutOpInterface, HasDefaultDLTIDataLayout, IsolatedFromAbove,
       SymbolTable, Symbol, SingleBlockImplicitTerminator<"ModuleEndOp">
     ]>, Arguments<(ins SymbolNameAttr:$sym_name,
           OptionalAttr<GPUNonEmptyTargetArrayAttr>:$targets,
-          OptionalAttr<OffloadingTranslationAttr>:$offloadingHandler)> {
+          OptionalAttr<OffloadingTranslationAttr>:$offloadingHandler,
+          OptionalAttr<SpatialExtentsAttr>:$spatialExtents)> {
```

Comment:
Isn't it a discardable (anonymous) attribute? I don't think having this in both places is a good idea.